csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,493 @@
+//! we traverse backwards, and we handle all active Expr outputs, this ensures
+//! that it is safe to removing an Expr will not result in erasing definitions
+//! of active Expr outputs.
+//!
+//! Derived classes should override handle() and make use of replaceTV(),
+//! markDeadAndMaybeRemove(), and allUsesDead(). Note that if replacements are
+//! made using replaceTV(old_tv, new_tv), then neither new_tv or any new
+//! Statements produced in creating it will be traversed by this class.
+class DeadCodeRemover : BackwardVisitor {
+ public:
+  DeadCodeRemover(Fusion* fusion) : BackwardVisitor(false), fusion_(fusion) {}
+
+  void run() {
+    // First we build a set of all live Statements so that we can detect dead
+    // branches.
+    auto exprs = StmtSort::getExprs(fusion_, fusion_->outputs());
+    // Mark every Expr, as well as its inputs and outputs as live initially
+    for (auto expr : exprs) {
+      markLive(expr);
+      for (auto inp : expr->inputs()) {
+        markLive(inp);
+      }
+      for (auto outp : expr->outputs()) {
+        markLive(outp);
+      }
+    }
+
+    // We do not traverse all outputs of all Exprs, since this requires that all
+    // paths lead to fusion_->outputs(). Instead, here we mark any Vals dead
+    // that do not have any live uses. After this, it is safe to check whether
+    // all outputs of an Expr are marked dead to determine if it should be dead,
+    // which helps us determine when it is safe to remove a multi-output
+    // definition of a dead TV.
+    std::vector<Statement*> dead_expr_outputs;
+    for (auto stmt : live_statements_) {
+      if (stmt->isVal() && !stmt->asVal()->isFusionOutput() &&
+          allUsesDead(stmt->asVal())) {
+        // We should not erase from live_statements_ while traversing it, so we
+        // save the dead expr outputs for later removal instead.
+        dead_expr_outputs.push_back(stmt);
+      }
+    }
+    for (auto stmt : dead_expr_outputs) {
+      markDead(stmt);
+    }
+
+    // Note that StmtSort::getExprs() is also run in traverseTo. In the future,
+    // we could potentially refactor this so that derived classes from
+    // BackwardVisitor can make use of that traversal instead of repeating it.
+    traverseTo(fusion_, fusion_->outputs());
+  }
+
+  Fusion* fusion() const {
+    return fusion_;
+  }
+
+ protected:
+  using BackwardVisitor::handle;
+
+  void handle(Statement* stmt) final {
+    if (!isLive(stmt)) {
+      // We check whether stmt is dead before we dereference it, since it may
+      // have been removed from the Fusion.
+      return;
+    }
+    BackwardVisitor::handle(stmt);
+  }
+
+  void handle(Expr* expr) final {
+    bool all_outputs_dead = true;
+    for (auto outp : expr->outputs()) {
+      if (isLive(outp)) {
+        all_outputs_dead = false;
+      }
+    }
+    if (all_outputs_dead) {
+      markDead(expr);
+      fusion_->removeExpr(expr);
```

Comment:
`BackwardVisitor::handle` does the dispatch for the `Statement`, but finding inputs and definitions for those objects is done earlier than this, in the body of `traverseTo`. Since it is done ahead of time, the sequence of `Statement`s to visit in `BackwardVisitor` is fixed, regardless of which changes are made to the `Fusion` during traversal. This includes removing `Statement`s from the `Fusion`, which is why we need to skip them via checking if they're dead. Some time after call, `expr`'s inputs will also be handled. They may not be dead themselves, but `removeExpr` removes `expr` from each input's `uses()` vector; even if it didn't, `expr` itself is marked dead. In any event, the use is either dead or gone entirely so the `allUsesDead()` check in `handle(TensorView* tv)` will mark the input dead if it actually should be.